### DISCUSSION TITLE
Question about Spatial Acceleration Structure Support

### DISCUSSION THREAD

**Discussion by moepforfreedom** [coelacant1/ProtoTracer#6] [Ideas]
Hello,

While looking through your code i noticed that the rendering loop currently checks all visible triangles sequentially when rendering a pixel. It would be possible to optimize this using a spatial acceleration structure (eg a Quad-/Octree, kd-tree or something similar), resulting in logarithmic runtime complexity. This typically results in a large performance improvement, especially for scenes with a large number of triangles. Is there any interest in implementing such an optimization in your project? If so then i could help with the implementation since i have implemented similar structures in my own raytracer projects.

**Comment by coelacant1** [comment]
> Hello,
> 
> While looking through your code i noticed that the rendering loop currently checks all visible triangles sequentially when rendering a pixel. It would be possible to optimize this using a spatial acceleration structure (eg a Quad-/Octree, kd-tree or something similar), resulting in logarithmic runtime complexity. This typically results in a large performance improvement, especially for scenes with a large number of triangles. Is there any interest in implementing such an optimization in your project? If so then i could help with the implementation since i have implemented similar structures in my own raytracer projects.

Hi Moep! Yes, this code is simply checking sequentially with basic bounds filtering for the triangles and certainly could have significant performance improvements using an alternative structure/design. I haven't gotten around to working on engine improvements like this in a while but it would be incredible to switch over to a spatial structure! If you're interested, then absolutely! Several users running on ESP32s could greatly benefit from this kind of calculation complexity!

**Reply by moepforfreedom** [reply]
Thats great to hear, i think as a start i could implement a simple screen-space quadtree that is generated after the animation/transform updates, this would only require minimal changes to the existing code. If thats fine for you then i can set up an implementation and submit a PR once its ready.
